File: Chapter12/python/structured_streaming/StreamStaticJoins.py
from pyspark.sql import SparkSession
from pyspark.sql import functions as F

spark = SparkSession \
    .builder \
    .appName("StructuredNetworkWordCount") \
    .getOrCreate()

# Create DataFrame representing the stream of input lines from connection to localhost:9999
lines = spark \
    .readStream \
    .format("socket") \
    .option("host", "localhost") \
    .option("port", 9999) \
    .load()

actions = spark.createDataFrame([(1, "opened website"), (2, "clicked"), (3, "scrolled"), (4, "left website")]) \
    .withColumnRenamed("_1", "actionID") \
    .withColumnRenamed("_2", "description")

webTraffic = lines.select(
    F.split("value", ",").getItem(0).alias("userID"),
    F.split("value", ",").getItem(1).alias("actionID")
)
enhanced = webTraffic.join(actions, "actionID")

# Generate running word count
actionCounts = enhanced.groupBy("description").count()

query = actionCounts \
    .writeStream \
    .outputMode("complete") \
    .format("console") \
    .start()

# Start running the query that prints the running counts to the console
query.awaitTermination()

File: Chapter12/structured_streaming/StreamStaticJoins.scala
package com.packtpub.spark.module_four.chapter_12.structured_streaming

import org.apache.spark.sql.SparkSession

object StreamStaticJoins {

  case class WebTraffic(userID: String, actionID: Int)
  case class Action(actionID: Int, description: String)

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .getOrCreate()

    import spark.implicits._

    val sc = spark.sparkContext
    sc.setLogLevel("WARN")

    // Establish Lookup Data for web traffic
    val actions = Seq((1, "opened website"), (2, "clicked"), (3, "scrolled"), (4, "left website"))
    val webActions =  spark.createDataset(actions).map(x => {
      Action(x._1, x._2)
    })

    // Make a DataFrame, `lines`, that will store every entry received from Netcat
    val lines = spark.readStream
      .format("socket")
      .option("host", "localhost")
      .option("port", 9999)
      .load()

    // Every line is just a comma-separated string received from the stream, so turn it into Dataset
    val webTrafficRaw = lines.as[String]

    val webTraffic = webTrafficRaw.map(x => {
      val parts = x.split(",")
      val userID = parts{0}
      val actionID = parts{1}.toInt
      WebTraffic(userID, actionID)
    }).join(webActions, "actionID")

    // Generate running word count
    val actionCounts = webTraffic.groupBy("description").count()

    // Start running the query that prints the running counts to the console
    val query = actionCounts.writeStream
      .outputMode("complete")
      .format("console")
      .start()

    query.awaitTermination()
  }
}
